### PR TITLE
feat: add `enforceForInnerExpressions` option to `no-extra-boolean-cast`

### LINKED ISSUE #18186: Rule Change: no-extra-boolean-cast should check final expression in comma operator expressions
A comma operator/sequence expression should have its last expression checked in a boolean context. For example, the following should flag:
```js
if ((1, 2, Boolean(3))) {
    // do something
}
```
since it is equivalent to 
```js
if ((1, 2, 3)) {
    // do something
}
```

The original issue description was significantly different from the decided upon course of action. It is preserved below for reference:

> ### What rule do you want to change?
>
> no-extra-boolean-cast
>
> ### What change do you want to make?
>
> Generate more warnings
> 
> ### How do you think the change should be implemented?
> 
> A new default behavior
> 
> ### Example code
> 
> ```js
> if (Boolean("don't flag") && Boolean("should flag!")) {
> }
> ```
>
>
> ### What does the rule currently do for this code?
> 
> by default: no error
> with { "enforceForLogicalOperands": true}: flags on both Boolean() calls
> 
> ### What will the rule do after it's changed?
> 
> By default: flag on `Boolean("should flag!")`
> with { "enforceForLogicalOperands": true}: same as current
> 
> ### Participation
> 
> - [X] I am willing to submit a pull request to implement this change.
> 
> ### Additional comments
> 
> This can be extended to the other "logical" operator varieties, i.e. `||` and `??`, and can also apply to "sequence expressions"/comma operator expressions (`if ((1, 2, Boolean(3)))`)

### PR DESCRIPTION
**Pull request by kirkwaiblinger**
Fixes #18186

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

fixes https://github.com/eslint/eslint/issues/18186

#### What changes did you make? (Give an overview)

~~Add support for checking "sequence expressions" ("comma operator") and right-hand-side operands of the nullish coalescing operator (`??`).~~

~~Note that neither is affected by the "enforceForLogicalOperands" option. The sequence expressions because it simply isn't one, and the `??` because it is not one either in this context. _(Yes, technically, the AST calls it a logical expression, but it's known to users as the [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing), and it really has nothing to do with logical operations at all, in the sense in which this rule applies to `&&` and `||`.)_~~

~~UPDATE - In addition to the cases mentioned in the issue (commas and `??`), I was reminded while working on this (as a result of other linter work I've done) that checking the branches of ternaries applies yet again for the same reasons. I've updated the PR with that change, but can easily revert it if it's deemed an overstep. (see the following examples for further examples of the same recursion pattern applying to commas, ternaries, optional chains, and "logical expressions": https://github.com/typescript-eslint/typescript-eslint/blob/59bbf417702b53fe78d8a190da7054bc6c22b780/packages/eslint-plugin/src/rules/no-floating-promises.ts#L221-L324 or https://github.com/typescript-eslint/typescript-eslint/blob/59bbf417702b53fe78d8a190da7054bc6c22b780/packages/eslint-plugin/src/rules/prefer-find.ts#L44-L114)~~

UPDATE 2 - After significant discussion in the PR, the approach is now to preserve existing behavior, deprecate `enforceForLogicalExpressions`, and create a new option (tentatively `enforceForInnerOperands`) that checks all of the recursive syntaxes: `&&`, `||`, RHS of `??`, right-most expression of sequence expressions `(a, b, c)`, and ternary consequent and alternate (`test ? consequent : alternate`)

### CODE REVIEW

**Comment by Tanujkanti4441** [comment]
Thanks for the PR, can you also add some examples in docs too?

File: lib/rules/no-extra-boolean-cast.js
Comment (mdjermanovic) [inline lib/rules/no-extra-boolean-cast.js:198]: This was assuming that `previousNode` is `parent.test` (as in the description of this function on line 167, which we should also update), but now it can be `consequent`/`alternate`, in which case only `SequenceExpression` would need parentheses.

For example:

```js
/* eslint no-extra-boolean-cast: 2 */

if (a ? Boolean(b = c) : Boolean(d = e));

```

After the autofix:

```js
/* eslint no-extra-boolean-cast: 2 */

if (a ? (b = c) : (d = e));

```

Parentheses around `b = c` and `d = e` are unnecessary.

Comment (kirkwaiblinger) [reply lib/rules/no-extra-boolean-cast.js]: This one took me a while to figure out trying to make use of the existing infrastructure. 
I would have expected to be able to write 
```js
                    if (previousNode === parent.test) {
                        return precedence(node) <= precedence(parent);
                    }
                    if (previousNode === parent.consequent || previousNode === parent.alternate) {
                        return precedence(node) < precedence(parent);
                    }
```
but it seems that eslint's precedence() function differs from MDN's table, in that it treats AssignmentExpressions and similar as lower precedence than ternary, rather than equal... See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_precedence#table, vs

https://github.com/eslint/eslint/blob/239a7e27209a6b861d634b3ef245ebbb805793a3/lib/rules/utils/ast-utils.js#L1509-L1602

Think that should be a followup to change to match the MDN version, or is it better to just leave the current system?

File: docs/src/rules/no-extra-boolean-cast.md
Comment (kirkwaiblinger) [inline docs/src/rules/no-extra-boolean-cast.md:65]: @Tanujkanti4441 Added some examples 👍

File: lib/rules/no-extra-boolean-cast.js
Comment (mdjermanovic) [reply lib/rules/no-extra-boolean-cast.js]: We can consider changing this in a followup if the current logic is confusing and the new one would be better for practical use, but we'll need to check all rules that use `getPrecedence()` to see if they need to be updated.

**Comment by mdjermanovic** [comment]
Looks good to me, but I'm not sure if we have a consensus on this change. Per the discussion in https://github.com/eslint/eslint/issues/18186, checking sequence expressions is accepted as a bug fix, but this also adds checking operands of `??` and `?:`.

I think it makes sense to check all these cases, but it looks unusual that checking operands of `||`/`&&` is behind an option, while checking operands of `??` and `?:` would be by default and without options.

**Comment by kirkwaiblinger** [comment]
> Looks good to me, but I'm not sure if we have a consensus on this change. Per the discussion in https://github.com/eslint/eslint/issues/18186, checking sequence expressions is accepted as a bug fix, but this also adds checking operands of `??` and `?:`.

Agreed. Lmk what a good course of action is. I can easily remove parts of the PR deemed out of scope (and, if requested, re-add them if a separate issue for them is accepted), or I can wait and watch for further discussion on whether to proceed with this PR as-is. I am unopinionated; making the code changes is easier than consensus-building :)

**Comment by Tanujkanti4441** [comment]
I think we can accept `??` and `?` operator as additional change in this PR. If we see following examples:
```js
if ((a, b, c ?? (d, e, f ?? Boolean(g)))) {}

if ((a, b, c ?? (d, e, f ?? Boolean(g)))) {}
```
then somehow it's a case of sequential expression and can be reported on `"enforceForLogicalOperands": true`.

**Comment by kirkwaiblinger** [comment]
> I think we can accept `??` and `?` operator as additional change in this PR. If we see following examples:
> 
> ```js
> if ((a, b, c ?? (d, e, f ?? Boolean(g)))) {}
> 
> if ((a, b, c ?? (d, e, f ?? Boolean(g)))) {}
> ```
> 
> then somehow it's a case of sequence expression and can be reported on `"enforceForLogicalOperands": true`.

Just to be clear, is there a specific action being requested here?

**Comment by Tanujkanti4441** [comment]
> Just to be clear, is there a specific action being requested here?

Not yet! it's just an opinion on reporting `??` and `?` operator when `"enforceForLogicalOperands": true` i think we need more feedback here

**Comment by mdjermanovic** [comment]
@eslint/eslint-team thoughts about https://github.com/eslint/eslint/pull/18222#issuecomment-2022705911?

**Comment by nzakas** [comment]
> I think it makes sense to check all these cases, but it looks unusual that checking operands of `||`/`&&` is behind an option, while checking operands of `??` and `?:` would be by default and without options.

I agree, this is a nice addition, but it does change this PR from a bug fix to a feature.

I also agree that it doesn't make sense to check `??` and `?` by default. I think it makes sense to add `?` to the `enforceForLogicalOperands` option, but I'm not sure about `??`.

**Comment by Rec0iL99** [comment]
ping @eslint/eslint-team. What are the next steps here?

**Comment by fasttime** [comment]
Sorry for the late reply. I agree that it makes sense to check the operands of `??` and `? :`. I don't have a strong opinion on whether those checks should be behind an option or not. If we want to put those checks behind an option to avoid reporting more problems unxpectedly, then I don't think we can reuse `enforceForLogicalOperands`, as that would in fact report more problems for configs that have this option set to `true`. If we just want to have the new checks behind an option because of the analogy with `||` and `&&`, then it's okay to add them to the cases already covered by `enforceForLogicalOperands`.

**Comment by nzakas** [comment]
@mdjermanovic thoughts on the feedback?

**Comment by mdjermanovic** [comment]
Looks like we all agree that it makes sense to check all these cases, and the only question is how the new checks would be enabled.

Since this is a stylistic rule, I don't think we can treat these as false negatives and thus enable the checks by default. I'd argue that the same applies to comma operands as well.

`enforceForLogicalOperands` also doesn't look like a good fit.

How about deprecating `enforceForLogicalOperands` and introducing a new option `enforceForInnerOperands` (or some better name)? The current behavior remains the same. When the new option is enabled, the rule checks all of `||`, `&&`, `??`, `?:`, `,`.

**Comment by nzakas** [comment]
> How about deprecating `enforceForLogicalOperands` and introducing a new option `enforceForInnerOperands` (or some better name)? The current behavior remains the same. When the new option is enabled, the rule checks all of `||`, `&&`, `??`, `?:`, `,`.

It looks like we are in agreement to take this approach. :+1:

**Comment by kirkwaiblinger** [comment]
@mdjermanovic @nzakas 

> How about deprecating enforceForLogicalOperands and introducing a new option enforceForInnerOperands (or some better name)? The current behavior remains the same. When the new option is enabled, the rule checks all of ||, &&, ??, ?:, ,.

I like this resolution!

As a suggestion, does anyone like the name `enforceForIndirectExpressions`?

**Comment by mdjermanovic** [comment]
> As a suggestion, does anyone like the name `enforceForIndirectExpressions`?

Sounds good to me, curious what others think.

**Comment by nzakas** [comment]
Works for me so long as we define "indirect expressions" fully in the docs.

**Comment by kirkwaiblinger** [comment]
I thought about it more and I decided I liked the word "inner" in @mdjermanovic's original suggestion, so I went with `enforceForInnerExpressions`. `indirect` was just bugging me lol. In any case, I think it should finally be good to go, now! 🙂

File: lib/rules/no-extra-boolean-cast.js
Comment (mdjermanovic) [inline lib/rules/no-extra-boolean-cast.js:40]: We could use `oneOf` to prevent specifying both options in the same config and thus avoid confusion (e.g., how is the rule expected to work when `enforceForInnerExpressions` is set to `true` and `enforceForLogicalOperands` explicitly set to `false`).

```js
schema: [{
    anyOf: [
        {
            type: "object",
            properties: {
                enforceForInnerExpressions: {
                    type: "boolean"
                }
            },
            additionalProperties: false
        },

        // depecated
        {
            type: "object",
            properties: {
                enforceForLogicalOperands: {
                    type: "boolean"
                }
            },
            additionalProperties: false
        }
    ]
}],
```

File: lib/rules/no-extra-boolean-cast.js
Comment (mdjermanovic) [inline lib/rules/no-extra-boolean-cast.js:131]: I think this whole function would be easier to follow if we now remove `function isLogicalContext` (which was a bit confusing name anyway) and move the condition here to align this check with the new checks below.

File: lib/rules/no-extra-boolean-cast.js
Comment (kirkwaiblinger) [reply lib/rules/no-extra-boolean-cast.js]: Good point! Changed!

File: docs/src/rules/no-extra-boolean-cast.md
Comment (mdjermanovic) [inline docs/src/rules/no-extra-boolean-cast.md:144]: Looks like this example is removed, can we leave it?

File: lib/rules/no-extra-boolean-cast.js
Comment (mdjermanovic) [inline lib/rules/no-extra-boolean-cast.js:227]: Can you also add one test where this returns `true`?

File: lib/rules/no-extra-boolean-cast.js
Comment (mdjermanovic) [inline lib/rules/no-extra-boolean-cast.js:159]: Can you also add one test where this is `false` (`node` isn't the last in the sequence expression)?

File: docs/src/rules/no-extra-boolean-cast.md
Comment (kirkwaiblinger) [reply docs/src/rules/no-extra-boolean-cast.md:144]: Good point. I added a few comments with it, too, since it's the only one that's not the result of a fix from the `incorrect` tab (which is probably why I forgot to include it when moving things around), and the reason that it's `correct` is not obvious at first glance

File: lib/rules/no-extra-boolean-cast.js
Comment (kirkwaiblinger) [reply lib/rules/no-extra-boolean-cast.js:227]: Good spot! - Added a test... The only thing that can be less is the comma operator, so I've added that.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! Since this is a new feature, it needs another approval from a team member before merging.

**Comment by mdjermanovic** [comment]
Thanks for contributing!

**Pull request merged**
